src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1914 @@
+}
+
+static int job_on_exit(sd_event_source *s, const siginfo_t *si, void *userdata) {
+        Job *j = ASSERT_PTR(userdata);
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *json = NULL;
+        int r;
+
+        assert(j);
+        assert(s);
+        assert(si);
+
+        if (IN_SET(j->type, JOB_UPDATE, JOB_VACUUM)) {
+                assert(j->target->busy);
+                j->target->busy = false;
+        }
+
+        if (si->si_code != CLD_EXITED) {
+                assert(IN_SET(si->si_code, CLD_KILLED, CLD_DUMPED));
+                sd_bus_error_setf(&error, SD_BUS_ERROR_FAILED,
+                                  "Job terminated abnormally with signal %s.",
+                                  signal_to_string(si->si_status));
+        } else if (si->si_status != EXIT_SUCCESS)
+                if (j->status_errno != 0)
+                        sd_bus_error_set_errno(&error, j->status_errno);
+                else
+                        sd_bus_error_setf(&error, SD_BUS_ERROR_FAILED,
+                                          "Job failed with exit code %i.", si->si_status);
+        else {
+                r = job_parse_child_output(TAKE_FD(j->stdout_fd), &json);
+                if (r < 0)
+                        sd_bus_error_set_errnof(&error, r, "Failed to parse JSON: %m");
+        }
+
+        /* Only send notification of exit if the job was actually detached */
+        if (j->detach_cb) {
+                r = sd_bus_emit_signal(
+                                j->manager->bus,
+                                "/org/freedesktop/sysupdate1",
+                                "org.freedesktop.sysupdate1.Manager",
+                                "JobRemoved",
+                                "toi",
+                                j->id,
+                                j->object_path,
+                                j->status_errno != 0 ? -j->status_errno : si->si_status);
+                if (r < 0)
+                        log_warning_errno(r, "Cannot emit JobRemoved message, ignoring: %m");
+        }
+
+        if (j->dbus_msg && j->complete_cb) {
+                if (sd_bus_error_is_set(&error)) {
+                        log_warning("Bus error occurred, ignoring callback for job: %s", error.message);
+                        sd_bus_reply_method_error(j->dbus_msg, &error);
+                } else {
+                        r = j->complete_cb(j->dbus_msg, j, json, &error);
+                        if (r < 0) {
+                                log_warning_errno(r, "Error during execution of job callback: %s", bus_error_message(&error, r));
+                                sd_bus_reply_method_errno(j->dbus_msg, r, &error);
+                        }
+                }
+        }
+
+        if (j->manager && hashmap_isempty(j->manager->jobs))
+                manager_became_idle(j->manager);
```

Comment:
I'd prefer if the `hashmap_isempty(j->manager->jobs)` check is moved inside manager_became_idle() (i.e. do an early exit here). Maybe even rename `manager_became_idle()` to `manager_check_idle()` or so.